src/azstoragetorch/io.py

```diff
@@ -0,0 +1,188 @@
+# license information.
+# --------------------------------------------------------------------------
+
+import io
+import os
+from typing import get_args, Optional, Union, Literal
+import urllib.parse
+
+from azure.identity import DefaultAzureCredential
+from azure.core.credentials import (
+    AzureSasCredential,
+    TokenCredential,
+)
+
+
+_SUPPORTED_MODES = Literal["rb"]
+_SDK_CREDENTIAL_TYPE = Optional[
+    Union[
+        AzureSasCredential,
+        TokenCredential,
+    ]
+]
+_AZSTORAGETORCH_CREDENTIAL_TYPE = Union[_SDK_CREDENTIAL_TYPE, Literal[False]]
+
+
+class BlobIO(io.IOBase):
+    def __init__(
+        self,
+        blob_url: str,
+        mode: _SUPPORTED_MODES,
+        *,
+        credential: _AZSTORAGETORCH_CREDENTIAL_TYPE = None,
+    ):
+        self._blob_url = blob_url
+        self._validate_mode(mode)
+        self._mode = mode
+        self._sdk_credential = self._get_sdk_credential(blob_url, credential)
+
+        self._blob_downloader = self._get_blob_downloader()
+        self._position = 0
+        self._closed = False
+
+    def close(self) -> None:
+        self._close_blob_downloader()
+        self._closed = True
+
+    @property
+    def closed(self) -> bool:
+        return self._closed
+
+    def fileno(self) -> int:
+        raise OSError("BlobIO object has no fileno")
+
+    def flush(self) -> None:
+        self._validate_not_closed()
+
+    def read(self, size: Optional[int] = -1, /) -> bytes:
+        if size is not None:
+            self._validate_is_integer("size", size)
+            self._validate_min("size", size, -1)
+        self._validate_not_closed()
+        return self._read(size)
+
+    def readable(self) -> bool:
+        self._validate_not_closed()
+        return self._mode == "rb"
+
+    def readline(self, size: Optional[int] = -1, /) -> bytes:
+        # BaseIO includes a default implementation of readline() that calls
+        # read() with a size of 1 until it finds a newline character. As of now, this has very poor
+        # performance implications as one read() will result in at least one HTTP GET request for each
+        # read() call.
+        #
+        # Instead of allowing the default implementation and having users impacted by these
+        # implications, we raise a NotImplementedError so that readline() and any other methods
+        # that are automatically implemented from readline() (e.g. readlines() and __next__()) can't
+        # be used until we implement a more appropriate strategy.
+        raise NotImplementedError("readline")
+
+    def seek(self, offset: int, whence: int = os.SEEK_SET, /) -> int:
+        self._validate_is_integer("offset", offset)
+        self._validate_is_integer("whence", whence)
+        self._validate_not_closed()
+        return self._seek(offset, whence)
+
+    def seekable(self) -> bool:
+        return self.readable()
+
+    def tell(self) -> int:
+        self._validate_not_closed()
```

Comment:
Same question here, is it typical for this to raise when the stream is closed?

Reply:
Yeah same thing here as in `readable()`; it's what the standard library file-like objects do:
```python
>>> import io
>>> f = io.BytesIO(b'foo')
>>> f.close()
>>> f.tell()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: I/O operation on closed file.
>>> with open('hello.txt', 'rb') as f:
...     pass
...
>>> f.tell()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: I/O operation on closed file
```
I think this behavior makes sense because if the file is closed, there is not really a current position in the file. So raising an error is better than returning a position that is not technically correct.